### ISSUE TITLE
Pass event object to onChange handler in toggle and select control

### ISSUE THREAD

**Issue by ravinderk** [WordPress/gutenberg#5506]
## Issue Overview

I currently explored only toggle and select control components and I found that onChange does not pass completed `event` object to handler instead just value.

Check to scenario:
```
const Inspector = ( props ) => {
	const attributes = props.attributes;

	// Event(s)
	const setDisplayStyleTo = displayStyle => {
		props.setAttributes( { displayStyle } );
	};

	const toggleShowTitle = () => {
		props.setAttributes( { showTitle: ! attributes.showTitle } );
	};

	const toggleShowGoal = () => {
		props.setAttributes( { showGoal: ! attributes.showGoal } );
	};

	const toggleContentDisplay = () => {
		props.setAttributes( { contentDisplay: ! attributes.contentDisplay } );

		// Set form Content Display Position
		if ( ! attributes.contentDisplay ) {
			props.setAttributes( { showContent: 'above' } ); // true && above
		} else if ( !! attributes.contentDisplay ) {
			props.setAttributes( { showContent: 'none' } ); // false && none
		}
	};

	const setShowContentPosition = position => {
		props.setAttributes( { showContent: position } );
	};

	const setContinueButtonTitle = continueButtonTitle => {
		props.setAttributes( { continueButtonTitle } );
		if ( ! props.isButtonTitleUpdated ) {
			props.updateButtonTitle( true );
		}
	};

	const updateContinueButtonTitle = () => {
		if ( props ) {
			props.doServerSideRender();
			props.updateButtonTitle( false );
		}
	};

	return (
		<InspectorControls key="inspector">
			<PanelBody title={ __( 'Display' ) }>
				<SelectControl
					label={ __( 'Form Format' ) }
					value={ props.attributes.displayStyle }
					options={ giveFormOptions.displayStyles }
					onChange={ setDisplayStyleTo } />
				{
					'reveal' === props.attributes.displayStyle && (
						<TextControl
							label={ __( 'Continue Button Title' ) }
							value={ props.attributes.continueButtonTitle }
							onChange={ setContinueButtonTitle }
							onBlur={ updateContinueButtonTitle } />
					)
				}
			</PanelBody>
			<PanelBody title={ __( 'Settings' ) }>
				<ToggleControl
					label={ __( 'Title' ) }
					checked={ !! props.attributes.showTitle }
					onChange={ toggleShowTitle } />
				<ToggleControl
					label={ __( 'Goal' ) }
					checked={ !! props.attributes.showGoal }
					onChange={ toggleShowGoal } />
				<ToggleControl
					label={ __( 'Content' ) }
					checked={ !! props.attributes.contentDisplay }
					onChange={ toggleContentDisplay } />
				{
					props.attributes.contentDisplay && (
						<SelectControl
							label={ __( 'Content Position' ) }
							value={ props.attributes.showContent }
							options={ giveFormOptions.contentPosition }
							onChange={ setShowContentPosition } />
					)
				}
			</PanelBody>
		</InspectorControls>
	);
};

export default Inspector;
```

In above code, I created multiple `onChange` handlers because `Gutenberg` API does not pass `event` object.
If we pass complete event object then I can reduce this code length and also get rid of multiple unwanted functions.

```
saveSetting(event) {
		const name = event.target.name;

		this.props.setAttributes(
			'checkbox' === event.target.type ?
				{ [name]: ! this.props.attributes[name] } :
				{ [name]: event.target.value }
		);
	}
```

Let me know if there is another way to handle this situation. Currently, I developed custom toggle and select control in which I am passing `event` object to `onChange` handler ( React also does same ).

**Comment by JimSchofield** [comment]
I encountered a similar issue in the PlaintText component.  I was expecting to have access to the event object, but I found it was bypassed in favor of just passing event.target.value to the callback. 

I suggest adding the event object as a second parameter for these common elements. This way we can still receive the event if we wish but we're not changing the current function of onChange.  Currently we need to rewrite components if we ever have need to access the event object.

`function PlainText( { onChange, className, ...props } ) {
	return (
		<TextareaAutosize
			className={ classnames( 'blocks-plain-text', className ) }
			onChange={ ( event ) => onChange( event.target.value, event ) }
			{ ...props }
		/>
	);
}`

**Comment by dmsnell** [comment]
Thanks for raising this question @ravinderk! I'd like to challenge us all to work within the constraints of not having DOM access (the event object). We can take a few different paths to pass data into our event handlers; one is the way you have created wrapper components and another is to pass closures/thunks as the event handlers…

```js
onChange={ value => onChange( 'showTitle', value ) }
```

that passes the information along without requiring us to create wrappers and go far out of our way.

**Comment by youknowriad** [comment]
Agreed with @dmsnell here. Ideally, we'd know why the event is needed and address these use-cases instead. A use-case here would be to share the same code between React and React Native and build environment agnostic blocks using these APIs.

Of course, it's still possible to build custom components with access to the event objects instead of using Gutenberg's provided components.

I'm closing this issue for now, we can reconsider later.

**Comment by M4N0JKUM4R** [comment]
I still don't get why we can't get the event object. I was trying to use [ColorPalette](https://developer.wordpress.org/block-editor/reference-guides/components/color-palette/) and then shifted towards using external react components like [react-color](https://casesandberg.github.io/react-color/) as they have the event object usage inbuilt. For example, let's say there are components generated through array map and each have their own index. How can I track it through the WordPress components way instead of using external React libraries?

I am currently doing like this using the external components, which seems easy but would want to use Gutenberg components as I don't want to mix it up.

```js
function handleBackgroundColor(color, event) {
    const parentIndex = event.target.closest(".legend-setting").dataset.index;
    
    console.log("Color", color);
    console.log("Event", event);
    
    console.log("Legend of the text", legend[parentIndex]);
    
    const newArray = [...legend];
    newArray[parentIndex].backgroundColor = color.hex;
    setLegend(newArray);
    setAttributes({ legend: newArray });
}
```

Return code:

```js
<div className="legend-settings-container" ref={legendContainerRef}>
	{legend.map((el, index) => (
		<>
	          <div className="legend-setting" data-index={index}>
			    <div className="background-color row">
				    <label>Legend Background Color</label>
				    
                                    {/* react-color way */}
				    <SketchPicker
					    color={el.backgroundColor}
					    onChangeComplete={(color, event) =>
						    handleBackgroundColor(color, event)
					    }
				    />
				    
                                   {/* ColorPalette way */}
				   <ColorPalette
                                          colors={ colors }
                                          value={ color }
                                          onChange={ ( color ) => setColor( color ) }
                                      />
			    </div>
		          </div>
	          </div>
		</>
	))}
</div>
```

**Comment by dmsnell** [comment]
Without speaking to why we _can_ or _can't_ get the object here's one way in which this can be accomplished through the WordPress component: we communicate from the parent to the child instead of from the child to the component. This is also a common idiom in a broad array of React-based designs as it plays well with React.

```js
const ParentComponent = ( props ) => {
    // do some stuff

    return (
        <SomeContainer>
            { props.legendItems.map( ( LegendItem, i ) => (
                <LegendItem onChangeBackground={ color => props.changeBackground( i, color ) } />
            ) }
        </SomeContainer>
    );
}
```

if taken to the extreme this leads to "prop drilling" which can be hard to follow, but then that's where state systems like Redux and MobX and others step in to separate the data updates from the UI hierarchy. the `updateLegendColor` function though would typically be given _to the child component_ from the parent which is rendering the list. While this does expand the API of the child components, it does let us avoid strongly coupling the implementation of the child to the context where it's rendered inside a list.

**Comment by M4N0JKUM4R** [comment]
@dmsnell Thanks for your time. Never expected a reply, so had some lazy edits in the first place :D If you see my fresh edit, 
I was probably thinking the events way instead of realizing that the map way of doing things, I rewrote my earlier code:

```js
<ColorPalette colors={colors} value={el.backgroundColor} onChange={(color) => handleBackgroundColor(index, color)} />
```

```js
function handleBackgroundColor(index, color) {
      const newArray = [...legend];
      newArray[index].backgroundColor = color;
      setLegend(newArray);
      setAttributes({ legend: newArray });
}
```

Easy thing that I missed.

But let's say if I were to get the target of the color component, do I have to use the `querySelector` way instead of the event object?

**Comment by dmsnell** [comment]
> But let's say if I were to get the target of the color component, do I have to use the querySelector way instead of the event object?

I'm not completely following what you wrote here. _If_ we get a DOM instance of something in the event we can access it, but on the other hand, if we can accomplish what we need without reaching into the DOM that generally has a number of advantages in React.

Dataflow is hard though, and keeping things more "pure" involves some complications, like passing these event handlers down from parent to child. One obvious perk of this design though is that we can easily test our component in isolation because we don't need to create a fake DOM environment and we don't have to create the surrounding DOM nodes that the child component would otherwise be looking for.

If you can be so kind as to try and describe your question differently I'll try and answer it differently 😄
